### PR TITLE
Updated main.py, minimaxchessplayer.py, and test_minimaxchessplayer.py.

### PR DESCRIPTION
**Pull request by WonkersTheWatilla**
In main.py: Updated MinimaxPlayer() args to include run_alpha_beta. This value can be changed to True or False depending on whether the players with use alpha-beta pruning or not. 

In minimaxchessplayer.py: changed 'min' and 'max' values in Nodes to alpha and beta. Changed default min and max values to ALPHA_DEFAULT and BETA_DEFAULT with corresponding infinity/-infinity values. Also added check for run_alpha_beta pruning in unpack() which calls alpha_beta_pruning() if True. Added an alpha_beta_pruning function that returns either the full array of successor states if no pruning occurred or a partial list of successor states if pruning has occurred. 

Updated test_minimaxchessplayer.py to match new MinimaxPlayer() arguments. Also added test to check if maximize is correctly identified and a second test to verify that alpha_beta_pruning returns a pruned set of successor moves [I just realized this needs to be fixed though].

To test the functionality of alpha-beta pruning I ran 3 tests against the generic minimax algorithm (with no heuristics) and the computation time was reduced by greater than 90% at depth=3. In addition, the results were still showing consistency. 

I continued to run a couple more tests at higher iterations [50, 500, 1000] and the results were still showing optimal game plays at a much faster computation time than minimax without alpha-beta-pruning.

### CODE REVIEW

**Comment by rfilgas** [comment]
Haven't been able to review alpha beta pruning fully yet, but given test results it looks reasonable so far. There are some files missed by the .gitignore that should be included, then of course the test you mentioned. Great work on this by the way!!

**Comment by WonkersTheWatilla** [comment]
Fixed the .gitignore issue. Ignore the many updates, the last one is the official one.

**Review by aferron — COMMENTED** [review]
I'm trusting your logic is working based on your test results

File: minimaxchessplayer.py [hunk 1]
```diff
@@ -151,20 +149,58 @@ def unpack(self,root: Node) -> np.array(Node):
...
+
+                if len(pruned_nodes) < len(nodes):
+                    return pruned_nodes
+                else:
+                    nodes = pruned_nodes
+
         return nodes
 
+    @staticmethod
+    def alpha_beta_pruning(maximizer: bool, value: int, children: np.array(Node), child: int) -> np.array(Node):
```
[the remark this answers is not in the document]
Comment (WonkersTheWatilla) [reply minimaxchessplayer.py]: Updated as requested.

File: minimaxchessplayer.py [hunk 2]
```diff
@@ -151,20 +149,58 @@ def unpack(self,root: Node) -> np.array(Node):
...
+            if self.run_alpha_beta:
+                # Check if child is a leaf
+                terminal_state = self.check_terminal_state(
+                    root=nodes[i],
+                    depth=depth,
+                    maximizer=true_root_board.turn
+                )
+                nodes[i].reward_if_taking_best_move = terminal_state if terminal_state is not None \
+                    else reward
+                pruned_nodes = self.alpha_beta_pruning(maximizer=maximizer, value=value, children=nodes, child=i)
```
[the remark this answers is not in the document]
Comment (WonkersTheWatilla) [reply minimaxchessplayer.py]: Updated as requested.

File: minimaxchessplayer.py [hunk 3]
```diff
@@ -151,20 +149,58 @@ def unpack(self,root: Node) -> np.array(Node):
...
+
+                if len(pruned_nodes) < len(nodes):
+                    return pruned_nodes
+                else:
```
[the remark this answers is not in the document]
Comment (WonkersTheWatilla) [reply minimaxchessplayer.py]: You are correct, that is unnecessary. I removed it.

File: minimaxchessplayer.py [hunk 4]
```diff
@@ -151,20 +149,58 @@ def unpack(self,root: Node) -> np.array(Node):
...
                 parent=root
             )
+
+            # If Alpha-beta-pruning, prune nodes
+            if self.run_alpha_beta:
+                # Check if child is a leaf
+                terminal_state = self.check_terminal_state(
+                    root=nodes[i],
+                    depth=depth,
+                    maximizer=true_root_board.turn
```
[the remark this answers is not in the document]
Comment (WonkersTheWatilla) [reply minimaxchessplayer.py]: Updated as requested.

File: minimaxchessplayer.py [hunk 5]
```diff
@@ -134,10 +128,14 @@ def pre_order(self, root_board: AIChessBoard, current: Node, depth: int) -> Node
...
 
     # Unpack nodes into their child states
-    def unpack(self,root: Node) -> np.array(Node):
+    def unpack(self, true_root_board: AIChessBoard, root: Node, depth: int) -> np.array(Node):
         base_board: AIChessBoard = root.board
         legalmoves = np.array(list(root.board.legal_moves))
         nodes = np.empty(len(legalmoves), dtype=Move)
+        # For alpha-beta-pruning
+        maximizer = True if (root.board.turn == true_root_board.turn) else False
+        value = ALPHA_DEFAULT if maximizer else BETA_DEFAULT
```
[the remark this answers is not in the document]
Comment (WonkersTheWatilla) [reply minimaxchessplayer.py]: Updated as requested.

File: minimaxchessplayer.py [hunk 6]
```diff
@@ -134,10 +128,14 @@ def pre_order(self, root_board: AIChessBoard, current: Node, depth: int) -> Node
 
 
     # Unpack nodes into their child states
-    def unpack(self,root: Node) -> np.array(Node):
+    def unpack(self, true_root_board: AIChessBoard, root: Node, depth: int) -> np.array(Node):
```
[the remark this answers is not in the document]
Comment (WonkersTheWatilla) [reply minimaxchessplayer.py]: Updated as requested.

File: minimaxchessplayer.py [hunk 7]
```diff
@@ -134,10 +128,14 @@ def pre_order(self, root_board: AIChessBoard, current: Node, depth: int) -> Node
...
         base_board: AIChessBoard = root.board
         legalmoves = np.array(list(root.board.legal_moves))
         nodes = np.empty(len(legalmoves), dtype=Move)
+        # For alpha-beta-pruning
+        maximizer = True if (root.board.turn == true_root_board.turn) else False
```
[the remark this answers is not in the document]
Comment (WonkersTheWatilla) [reply minimaxchessplayer.py]: Updated as requested.

File: minimaxchessplayer.py [hunk 8]
```diff
@@ -151,20 +149,58 @@ def unpack(self,root: Node) -> np.array(Node):
...
+                pruned_nodes = self.alpha_beta_pruning(maximizer=maximizer, value=value, children=nodes, child=i)
+
+                if len(pruned_nodes) < len(nodes):
+                    return pruned_nodes
+                else:
+                    nodes = pruned_nodes
+
         return nodes
 
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (WonkersTheWatilla) [reply minimaxchessplayer.py]: No longer a static method.

**Review by rfilgas — DISMISSED** [review]

**Comment by WonkersTheWatilla** [comment]
Updated main.py and minimax.py to match new merges.

**Review by aferron — APPROVED** [review]
Great job

**Pull request merged**